src/common/substrait/src/df_expr.rs

```diff
@@ -0,0 +1,714 @@
+};
+
+/// Convert substrait's `Expression` to DataFusion's `Expr`.
+pub fn to_df_expr(ctx: &ConvertorContext, expression: Expression, schema: &Schema) -> Result<Expr> {
+    let expr_rex_type = expression.rex_type.context(EmptyExprSnafu)?;
+    match expr_rex_type {
+        RexType::Literal(_) => todo!(),
+        RexType::Selection(selection) => convert_selection_rex(selection, schema),
+        RexType::ScalarFunction(scalar_fn) => convert_scalar_function(ctx, scalar_fn, schema),
+        RexType::WindowFunction(_)
+        | RexType::IfThen(_)
+        | RexType::SwitchExpression(_)
+        | RexType::SingularOrList(_)
+        | RexType::MultiOrList(_)
+        | RexType::Cast(_)
+        | RexType::Subquery(_)
+        | RexType::Enum(_) => UnsupportedExprSnafu {
+            name: format!("substrait expression {:?}", expr_rex_type),
+        }
+        .fail()?,
+    }
+}
+
+/// Convert Substrait's `FieldReference` - `DirectReference` - `StructField` to Datafusion's
+/// `Column` expr.
+pub fn convert_selection_rex(selection: Box<FieldReference>, schema: &Schema) -> Result<Expr> {
+    if let Some(FieldReferenceType::DirectReference(direct_ref)) = selection.reference_type
+    && let Some(SegReferenceType::StructField(field))  = direct_ref.reference_type{
+        let column_name = schema.column_name_by_index(field.field as _).to_string();
+         Ok(Expr::Column(Column{ relation: None, name: column_name }))
+    } else{
+        InvalidParametersSnafu{reason:"Only support direct struct reference in Selection Rex"}.fail()
+    }
+}
+
+pub fn convert_scalar_function(
+    ctx: &ConvertorContext,
+    scalar_fn: ScalarFunction,
+    schema: &Schema,
+) -> Result<Expr> {
+    // convert argument
+    let mut inputs = VecDeque::with_capacity(scalar_fn.arguments.len());
+    for arg in scalar_fn.arguments {
+        if let Some(ArgType::Value(sub_expr)) = arg.arg_type {
+            inputs.push_back(to_df_expr(ctx, sub_expr, schema)?);
+        } else {
+            InvalidParametersSnafu {
+                reason: "Only value expression arg is supported to be function argument",
+            }
+            .fail()?;
+        }
+    }
+
+    // convert this scalar function
+    // map function name
+    let anchor = scalar_fn.function_reference;
+    let fn_name = ctx.find_scalar_fn(anchor).context(InvalidParametersSnafu {
+        reason: format!("Unregistered scalar function reference: {}", anchor),
+    })?;
+
+    // convenient util
+    let ensure_arg_len = |expected: usize| -> Result<()> {
+        ensure!(
+            inputs.len() == expected,
+            InvalidParametersSnafu {
+                reason: format!(
+                    "Invalid number of scalar function {}, expected {} but found {}",
+                    fn_name,
+                    expected,
+                    inputs.len()
+                )
+            }
+        );
+        Ok(())
+    };
+
+    // onstruct DataFusion expr
+    let expr = match fn_name {
```

Comment:
maybe we need to maintain a document about the mapping rules 🫠 I should investigate how to document it properly.